Rails 7 update

amygurski opened the pull request:
### What github issue is this PR for, if any?
Resolves #3283 

### What changed, and why?
Bump rails from 6.1.4.6 to 7.0.2.3. Run the rails update task and change the configuration defaults to use rails 7.0 defaults accordingly.

### How will this affect user permissions?
- Volunteer permissions: N/A
- Supervisor permissions: N/A
- Admin permissions: N/A

### How is this tested? (please write tests!) 💖💪
No new tests since no new functionality, just make sure existing functionality unchanged.
- [X] Ensure app boots up cleanly; no failures or depreciation warnings in logs on startup.
- [X] Manual UI testing: 
  - Click around site - creating, editing and deleting records. 
  - Pay particular attention to the `button_to` elements since the behavior there has changed slightly. 
  - `bullet` is logging lots of warnings locally for AVOID/USE eager loading, but I confirmed they were there before the upgrade.
- [X] `rails console` works (a known rails 7 + old spring issue)
- [X] `rspec`: investigate and address failures in existing tests due to rails 7 changes; run and watch for any **new** warnings ignoring pre-existing warnings.

### Screenshots please :)


### Feelings gif (optional)
What gif best describes your feeling working on this issue? https://giphy.com/
How to embed:
![simpsons-spring-break](https://media.giphy.com/media/3orifiPHeQc4rkKGg8/giphy.gif)

### Feedback please? (optional)
We are very interested in your feedback! Please give us some :) https://forms.gle/1D5ACNgTs2u9gSdh9

amygurski committed:
Bump rails from 6.1.4.6 to 7.0.2.3

amygurski committed:
Apply rails 7 config changes from rails update task

amygurski committed:
Remove spring-watcher-listen gem

spring-watcher-listen doesn't allow for spring v3.0 currently but
spring v3.0 is mandatory to upgrade to rails 7. Otherwise, it's not
possible to use spring or load a rails console.

This can be added back in later if desired, but it's unlikely to be a
noticable change. Spring will use the default way of watching the filesystem
for changes without this gem. For this size app and modern computers
it's unlikely that there  will be any noticable difference in responsiveness
or cpu use on our local systems during development.

If we want to re-add it, it can be done once this issue is resolved:
https://github.com/rails/spring-watcher-listen/issues/27

amygurski committed:
Bump spring from 2.1.1 to 4.0.0

spring needs to be at least 3.0.0 to be compatible with rails 7:
https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-6-1-to-rails-7-0-spring

Prior to this upgrade, a rails console failed to open with error:
undefined method 'mechanism=' for ActiveSupport::Dependencies:Module

amygurski committed:
Use default file watcher and remove listen gem

The rails update task for rails 7 removed the config for
ActiveSupport::EventedFileUpdateChecker which also required the listen
gem. It turns out they no longer recommend this as the default because
no benefits are seen so may as well simplify things.

Reference:
https://github.com/rails/rails/commit/e34300a921a910c60cd4064038b0817fe0e0ab18

amygurski committed:
Load rails 7.0 configuration defaults

Load all rails 7 defaults except button_to_generates_button_tag.
This caused failures with capybara being unable to find the css on the
"Unassign Volunteer" button in spec/system/casa_cases/edit_spec.

amygurski committed:
Keep using imagemagick in rails 7

Rails changed their default active storage variant processor from
imagemagick to vips in rails 7. This caused the casa site to crash when manually
testing because of the vips library being unavailable. According to their docs,
it is fine to keep using imagemagick and this option is not being deprecated
so chose to keep it rather than replacing it.

https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#active-storage-default-variant-processor-changed-to-vips

amygurski committed:
Fix unsafe redirects in tests using example.com

Rails 7 protects from unsafe redirects by default now. Change the tests
to use the root path as the http referrer instead of example.com so it's
allowed to fix the tests. Didn't modify the controller to allow unsafe
redirects because I'm assuming we expect the referrer when a court
date or case contact is deleted to be a URL within our site.

compwron reviewed (commented):
This is looking great! :)

amygurski committed:
Enable default button_to_generates_button_tag config

config.action_view.button_to_generates_button_tag = true is a default
configuration in rails 7. It ensures that the button_to helper method
consistently renders a button element instead of an input element.

This page summarizes this change succinctly:
https://blog.saeloun.com/2021/08/24/rails-7-button-to-rendering.html
